Fix phpcomplete for patches 8.2.2426 and 8.2.2427

complex857 opened the pull request:
Changing windows while inside a completion function is forbidden now.
Migrated to use popup windows instead of split windows.

Special thanks to @Roy-Orbison for reporting it and @lacygoill for the
popup window suggestion

Fixes #7783

Commit:
Fix plugin for patches 8.2.2426 and 8.2.2427

Changing windows while inside a completion function is forbidden now.
Migrated to use popup windows instead of split windows.

Special thanks to @Roy-Orbison for reporting it and @lacygoill for the
popup window suggestion

Fixes #7783

lacygoill commented on runtime/autoload/phpcomplete.vim:
```diff
@@ -2082,26 +2078,19 @@ function! phpcomplete#GetClassContentsStructure(file_path, file_lines, class_nam
-
-	let content = join(getline(cfline, endline), "\n")
+	let popup_id = popup_create(a:file_lines, {'hidden': v:true})
+    call win_execute(popup_id, 'syntax on')
```
Executing `:syntax on` on every completion looks inefficient, and might have undesirable effects.  And I don't think it needs to be executed from `win_execute()` since its effect should not depend on the current window.  Also, if the syntax was not enabled, it should be disabled at the end.  So, instead, you could write this (right from `phpcomplete#GetClassContentsStructure()`; no need to put this inside `win_execute()`):

    let syntax_was_enabled = exists('g:syntax_on')
    if !syntax_was_enabled
        syntax enable
    endif

    " do the stuff which requires the syntax to be enabled

    if !syntax_was_enabled
        syntax off
    endif

---

And if you need some syntax highlighting to be applied in your popup, for example `php`, then you can execute this:

    call win_execute(popup_id, 'do Syntax php')

Or maybe this:

    call win_execute(popup_id, 'set syntax=php')

lacygoill commented on runtime/autoload/phpcomplete.vim:
```diff
@@ -2532,40 +2521,37 @@ function! phpcomplete#FormatDocBlock(info) " {{{
-	silent! 0put =a:file_lines
-	normal! G
+	let popup_id = popup_create(a:file_lines, {'hidden': v:true})
+	call win_execute(popup_id, 'syntax on')
```
Same remark as previously.

Roy-Orbison commented:
While you're there, I also noticed what looks like a stray file read command (into an unused variable in that scope) which executes before the cache is read. If I'm right, it can't be good for performance. I did run a test by removing that line and doing a completion (in Vim 7.4) and nothing seemed to break. Could you review and add that to this PR?

https://github.com/complex857/vim/blob/8add49ae49dfcf5570d0d16b1d80cb636e387bbb/runtime/autoload/phpcomplete.vim#L183

Roy-Orbison commented:
This doesn't fully solve the problem for me. I've tried both this version as well as the one in the shawncplus repo, and both seem to throw an exception at [a particular `win_execute()`](https://github.com/complex857/vim/blob/8add49ae49dfcf5570d0d16b1d80cb636e387bbb/runtime/autoload/phpcomplete.vim#L2543) line.

My debugger log shows this:
```text
>step
line 60: let [current_namespace, imports] = phpcomplete#GetCurrentNameSpace(getline(0, line('.')))
line 1: let popup_id = popup_create(a:file_lines, {'hidden': v:true})
line 2: call win_execute(popup_id, 'syntax on')
line 3: call win_execute(popup_id, 'set ft=phpcompletetempbuffer')
line 4: call win_execute(popup_id, 'normal! G')
line 7: while 1
line 8: call win_execute(popup_id, 'let block_start_pos = searchpos(''\c\(class\|trait\|function\|interface\)\s\+\_.\{-}\zs{'', "Web")')
line 9: if block_start_pos == [0, 0]
line 11: endif
line 12: call win_execute(popup_id, 'let block_end_pos = searchpairpos("{", "", ''}\|\%$'', "W", ''synIDattr(synID(line("."), col("."), 0), "name") =~? "string\\|comment"'')')
line 14: if block_end_pos != [0, 0]
line 16: call win_execute(popup_id, block_start_pos[0].','.block_end_pos[0].'d _')
>echo block_start_pos[0].','.block_end_pos[0].'d _'
73,76d _
line 16: d _
>step
Exception thrown: Vim(delete):E578: Not allowed to change text here
     function phpcomplete#CompletePHP[60]..phpcomplete#GetCurrentNameSpace
line 16: End of function
```

I'm also not great at vim script but it looks like there're unnecessary `exec`s in there since both `:execute` and `win_execute()` both run commands in Ex mode according to the docs. I changed that section to the following:
```viml
		if block_end_pos != [0, 0]
			" end of the block found, just delete it
			call win_execute(popup_id, block_start_pos[0].','.block_end_pos[0].'d _')
		else
			" block pair not found, use block start as beginning and the end
			" of the buffer instead
			call win_execute(popup_id, block_start_pos[0].',$d _')
		endif
```
Unfortunately, it did not resolve the error, but nor did it change.

lacygoill commented:
> I'm also not great at vim script but it looks like there're unnecessary execs in there since both :execute and win_execute() both run commands in Ex mode according to the docs.

Yes, there's no need to use `:exe` with `win_execute()`.

> This doesn't fully solve the problem for me.

What is the contents of the php file on which you experience the issue?

Roy-Orbison commented:
Without debugging mode I get a red and white error message that says:
```text
Error detected while processing function phpcomplete#CompletePHP[127]..function phpcomplete#CompletePHP[60]..phpcomplete#GetCurrentNameSpace:
```
It doesn't seem to matter what the source is in my project, it keeps failing.

If I use a minimal example (separate tags file for a tiny, dummy class: `Foo` with method `bar`, run `vim -u NONE -N`, only `:source` the fixed completion script, and a single line in a new buffer to complete like `Foo::ba`) then the error does not occur (but does say no matches for the form `(new Foo)->ba`).

If I run the same minimal vim inside my project, and do a basic class completion, it fails with the above error and I can't figure out why. The only real difference is the tags file, yet I use the same program & invocation to generate the tags files for the project and the `Foo` test.

lacygoill commented:
I don't think that deleting lines in a popup should raise `E578`; it's weird that it does.  Maybe we could work around that by using `popup_settext()`.

If you can find a minimal reproducible example (vimrc, php file, tag file if necessary), I will try to find a fix.

Commit:
Remove syntax manipulation on popup windows

This was not part of the distribution version of the plugin, copied over
by mistake here.

Roy-Orbison commented:
> If you can find a minimal reproducible example (vimrc, php file, tag file if necessary), I will try to find a fix.

I've been doing my best :) I shall continue trying.

lacygoill commented:
I still can't reproduce with the plugin, but I think I've found another example of this issue:

    vim -Nu NONE -S <(cat <<'EOF'
        vim9
        nno <expr> <F3> <sid>Func()
        def Func()
            var id = popup_create(['aaa', 'bbb', 'ccc'], {hidden: true})
            win_execute(id, ':2d')
        enddef
        Func()
        feedkeys("\<F3>")
    EOF
    )

I have a few ideas to work around this limitation.

complex857 commented:
> I also noticed what looks like a stray file read command

You're right @Roy-Orbison I've removed that one

At first i've had those lines without the `exec`, but I was getting invalid command errors.
I think i've a basic setup to reproduce them, needs 3 files:
```
// t2.php
<?php

$f = new Foo;
$f-> // <--- starting completion here
```

```
// t3.php
<?php

class Foo {
  public function find_me() {
  }
}
```
tag file: 
```
!_TAG_FILE_FORMAT	2	/extended format; --format=1 will not append ;" to lines/
!_TAG_FILE_SORTED	1	/0=unsorted, 1=sorted, 2=foldcase/
!_TAG_OUTPUT_MODE	e-ctags	/u-ctags or e-ctags/
!_TAG_PROGRAM_AUTHOR	Universal Ctags Team	//
!_TAG_PROGRAM_NAME	Universal Ctags	/Derived from Exuberant Ctags/
!_TAG_PROGRAM_URL	https://ctags.io/	/official site/
!_TAG_PROGRAM_VERSION	0.0.0	/509fa4f/
Foo	t3.php	/^class Foo {$/;"	c	language:PHP
f	t2.php	/^$f = new Foo;$/;"	v	language:PHP
find_me	t3.php	/^  public function find_me() {$/;"	f	language:PHP	class:Foo	access:public	signature:()
```

After some experimenting it seems that the following change makes the `win_exec()` do what it should here:

```diff
-	call win_execute(popup_id, 'exec "'.block_start_pos[0].','.block_end_pos[0].'d _"')
+	call win_execute(popup_id, '"'.block_start_pos[0].','.block_end_pos[0].'d _"')
 else
 	" block pair not found, use block start as beginning and the end
 	" of the buffer instead
-	call win_execute(popup_id, 'exec "'.block_start_pos[0].',$d _"')
+	call win_execute(popup_id, '"'.block_start_pos[0].',$d _"')
```
Tried it without the leading `'"'.` and that generated the same error as you see (same with the `exec ...` in front of it).
It's a bit of a mystery for me why, I've to say 😅 

@Roy-Orbison could you confirm if those changes make the issue go away on your version?

lacygoill commented:
I could be wrong, but I don't think this fix is correct, because the leading double quote causes the command to be disabled (commented).  So, yes, it will prevent the error from being raised, but that's because it no longer does anything.

If I'm wrong and it works correctly, then that's fine.  Otherwise, I'll post a solution later.  I can't do it right now, because in the process of finding a workaround I found a Vim bug which I need to report.

complex857 commented:
Ah, indeed you are right! It turns itself into a comment, forgot the viml syntax here 🤦

The deletion is not 100% necessary here (removes bodies of functions, so latter commands run a bit faster), that's why turning it into a noop seem to have "worked"

Roy-Orbison commented:
> You're right @Roy-Orbison I've removed that one

Cool.

> I think i've a basic setup to reproduce them, needs 3 files:

That was essentially what I had, too.

> @Roy-Orbison could you confirm if those changes make the issue go away on your version?

Sorry, I tried your changes, both with double and single quotes. I didn't get an error, but nor did any methods autocomplete. Thanks for trying.

lacygoill commented:
It find it weird that Vim complains when we try to delete lines in a popup while a completion is active.  I could understand that it complains about lines in the current buffer, but not in a popup buffer.  Maybe making the distinction between the 2 is not that easy (e.g. the popup could display the current buffer); I don't know.

Anyway, one solution is to get the lines in the popup with `getbufline()`, remove the undesirable lines with `remove()`, then reset the popup contents with `popup_settext()`.  In effect, this should be equivalent to `:delete`.  To get the buffer number, `winbufnr()` can be used (it needs the popup window id as argument).

Once I'm done with the Vim bug, I'll try to post a patch, unless someone else does it before.

lacygoill commented:
Can someone try this patch?
```diff
diff --git a/runtime/autoload/phpcomplete.vim b/runtime/autoload/phpcomplete.vim
index aa70bc12e..ced033ccf 100644
--- a/runtime/autoload/phpcomplete.vim
+++ b/runtime/autoload/phpcomplete.vim
@@ -2533,14 +2533,16 @@ function! phpcomplete#GetCurrentNameSpace(file_lines) " {{{
 		endif
 		call win_execute(popup_id, 'let block_end_pos = searchpairpos("{", "", ''}\|\%$'', "W", ''synIDattr(synID(line("."), col("."), 0), "name") =~? "string\\|comment"'')')
 
+		let popup_lines = winbufnr(popup_id)
 		if block_end_pos != [0, 0]
 			" end of the block found, just delete it
-			call win_execute(popup_id, 'exec "'.block_start_pos[0].','.block_end_pos[0].'d _"')
+			call remove(popup_lines, block_start_pos[0] - 1, block_end_pos[0] - 1)
 		else
 			" block pair not found, use block start as beginning and the end
 			" of the buffer instead
-			call win_execute(popup_id, 'exec "'.block_start_pos[0].',$d _"')
+			call remove(popup_lines, block_start_pos[0] - 1, -1)
 		endif
+		call popup_settext(popup_id, popup_lines)
 	endwhile
 	call win_execute(popup_id, 'normal! G', 'silent!')
 
```

lacygoill commented:
Ah no the patch is wrong.  Let me fix it.

lacygoill commented:
New patch:
```diff
diff --git a/runtime/autoload/phpcomplete.vim b/runtime/autoload/phpcomplete.vim
index aa70bc12e..ced033ccf 100644
--- a/runtime/autoload/phpcomplete.vim
+++ b/runtime/autoload/phpcomplete.vim
@@ -2533,14 +2533,16 @@ function! phpcomplete#GetCurrentNameSpace(file_lines) " {{{
 		endif
 		call win_execute(popup_id, 'let block_end_pos = searchpairpos("{", "", ''}\|\%$'', "W", ''synIDattr(synID(line("."), col("."), 0), "name") =~? "string\\|comment"'')')
 
+		let popup_lines = winbufnr(popup_id)->getbufline(1, '$')
 		if block_end_pos != [0, 0]
 			" end of the block found, just delete it
-			call win_execute(popup_id, 'exec "'.block_start_pos[0].','.block_end_pos[0].'d _"')
+			call remove(popup_lines, block_start_pos[0] - 1, block_end_pos[0] - 1)
 		else
 			" block pair not found, use block start as beginning and the end
 			" of the buffer instead
-			call win_execute(popup_id, 'exec "'.block_start_pos[0].',$d _"')
+			call remove(popup_lines, block_start_pos[0] - 1, -1)
 		endif
+		call popup_settext(popup_id, popup_lines)
 	endwhile
 	call win_execute(popup_id, 'normal! G', 'silent!')

```

Roy-Orbison commented:
@lacygoill That's done the trick! Hooray! :champagne:

Thanks to both of you for prompt refactoring of the code. It's also now completing the `(new Foo)->ba` format partial names again.

Commit:
Patch from @lacygoill to avoid "Not allowed to change text here" errors

brammool commented:
Thanks all,  I'll include it now. Will be in the next runtime files update.
